### ISSUE TITLE
DL4J 1.0.0-M1 no longer compatible with older stdlibc++

### ISSUE THREAD

**Issue by reckart** [deeplearning4j/deeplearning4j#9352]
#### Issue Description

While trying to upgrade from 1.0.0-beta7 to M1, we hit the following issue on our CI server:

```
Caused by: java.lang.UnsatisfiedLinkError: /var/lib/jenkins/.javacpp/cache/nd4j-native-1.0.0-M1-linux-x86_64.jar/org/nd4j/nativeblas/linux-x86_64/libjnind4jcpu.so: /lib64/libstdc++.so.6: version `CXXABI_1.3.8' not found (required by /var/lib/jenkins/.javacpp/cache/nd4j-native-1.0.0-M1-linux-x86_64.jar/org/nd4j/nativeblas/linux-x86_64/libjnind4jcpu.so)
```

It appears as if the milestone release used a different (newer) g++ compiler version than the previous beta versions to build the some native libraries, which makes it incompatible with hosts running older Linux versions (e.g. CentOS 7).

Was this an intentional breaking change?

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version: 1.0.0-M1
* Platform information: CentOS 7

**Comment by agibsonccc** [comment]
@reckart we migrated everything to github actions using the oldest ubuntu available 16.04. What is your production system configuration?  Is it just the CI server? We will be doing a quick follow up release to address user concerns from the migration. Depending on the system you use though, we may have to figure out an alternative for you.

**Comment by agibsonccc** [comment]
Full updates for centos 7 just ended in q4 of last year. The trade off was made for the migration to cut down on maintence costs.  The new docs will cover this: https://github.com/KonduitAI/deeplearning4j-gitbook-sync/blob/en-1.0.0-m1/multi-project/how-to-guides/developer-docs/github-actions-overview.md

**Comment by agibsonccc** [comment]
@reckart  feel free to dm me on the community forums if you have something in specific you can't discuss in public and we can also see a good way to help your team have a smoother time with the migration.

**Comment by reckart** [comment]
@agibsonccc If it was an intentional breaking change, then so it is. We have been exploring a bunch of options on our side how to deal with the issue (e.g. updating the system, building in Docker, even rolling back the DL4J update) - but before making a decision on how we should proceed, I wanted to verify that this change is indeed intentional and not something that somehow slipped through QA and might be reverted.

**Comment by agibsonccc** [comment]
@reckart  I can explore adding centos 7 for our cpu builds. Github actions has a way of running container builds. Let me see what adding an alternative looks like. Ubuntu 16 was fine for some people.  Given that glibc is a fairly consistent issue for people, I may do multiple spins. This would come up if you wanted to leverage avx extensions for example (only newer gcc supports it) - so there's always a trade off of compatibility, performance and compiled native code. 

Could you elaborate a bit on your use case? It certainly wouldn't hurt to  let users be able to benefit from performance improvements and the like. I'm open to working with the community to find something suitable. Now that the builds are setup, adding another one would be some work, but should be do able.

**Comment by reckart** [comment]
@agibsonccc The use-case is pretty boring, I'm afraid. Some of the machines we use are a bit dated and so is the OS on them (patched to the max of course). Some even predate AVX2. We have already observed that these can no longer be used with the recent tensorflow  Java API which no longer ships non-AVX2 versions. But we didn't hit the ABI issue with tensorflow (yet) - only with DL4J. We tend to be conservative with dropping support for old versions of anything (hardware and software) to avoid running into a situation where deployment at a client would entail a hardware upgrade. For the software, deployment is facilitated via docker. For our CI setup, using docker is a bit trickier because of the interaction between the main build, "sidecar" docker containers running services, the CI server and the Maven-based build.

I was happily surprised when I recently read that the new DL4J versions would be capable of automatically choosing between AVX2 and non-AVX libraries at load-time instead (at least that's what I understood from comments on the respective issue at the time). 

So for the time being, it would be convenient for us if "legacy" CPUs and OS versions would still be supported - but to be honest, I also know that their days are pretty much counted and one can't support the old stuff forever. 

So I opened the issue on the notion that maybe you'd not be aware of the issue (yet) and might from your vantage point figure that a relevant portion of your user-base might be running into the same issue and make a change --- or that you intentionally decide to go forward with the situation as it is now and let us know, so that we can factor that into planning our next steps.

**Comment by agibsonccc** [comment]
@reckart super useful thank you! We've had a few hits for the libc issue, but most people have been happy with ubuntu 16's glibc. I'll evaluate what might be possible in the next week or so as we finish the planning for the bug fix release going out soon. In github actions, you *can* specify a container. It's good to know what the baseline is, and the feedback is super helpful.

**Comment by agibsonccc** [comment]
@reckart to keep in line with older glibcs, we'll add a compat classifier to handle this case. It will have to come after our release, but it's worth doing. I think having a newer default will be better with people forced to evaluate a fallback for older releases. It will also help educate users on the trade offs of linking against newer versions. Otherwise I don't think a lot of folks will be aware of our newer classifiers like ondnn or avx.

**Comment by reckart** [comment]
Wrt: compatibility with older OSes - great!

Wrt classifiers: to the different variations of the native libs (i.e. classifiers). I believe by default when somebody uses DL4J, all variations of native libraries in principle available and at runtime, DL4J should inspect the environment (i.e. CPU flags, OS capabilities) and decide which variation is the best to use on that particular system). At least in beta6, if there is AVX and I am using a non-AVX, I get a message to change to the AVX libraries. But this is a system-specific choice. If I deploy to my customer's systems, they may not have AVX. So to avoid having to create multiple versions of a product, it would be better if libraries like DL4J that have native parts could automatically choose the right one - not only for the particular OS they are running on, but also depending on circumstances such as CPU flags, OS capabilities or actually even the presence and absence of GPUs. 

That said, if I want to trim down the dependency footprint of my product, it would also be great if there was a convenient way of doing so. Maybe there would be a way do support something like:

```
<dependency>
  <groupId>org.eclipse.dl4j<groupId>
  <artifactId>dl4j</artifactId>
  <excludes>
    <exclude>
      <groupId>*<groupId>
      <artifactId>*</artifactId>
      <classifier>android</classifier>
    </exclude>
  </excludes>
</dependency>
```

I didn't try it out, but I think the classifier may not be supported in excludes - but maybe an artifactId suffix could be used instead of a classifier.

Or alternatively have convenience BOMs like:

```
<dependency>
  <groupId>org.eclipse.dl4j<groupId>
  <artifactId>dl4j-bom-x86-compat</artifactId>
</dependency>
<dependency>
  <groupId>org.eclipse.dl4j<groupId>
  <artifactId>dl4j-bom-x86-avx2</artifactId>
</dependency>
<dependency>
  <groupId>org.eclipse.dl4j<groupId>
  <artifactId>dl4j-bom-x86-dnn</artifactId>
</dependency>
```

tl;dr: include everything by default, choose at runtime (or rather at start-up time - not expecting dynamic switching), give users an easy way to tune dependency footprint when desired

**Comment by agibsonccc** [comment]
@reckart  the main problem is we compile everything ahead of time. Basically we do libnd4j c++ -> put that in to jar file for javacpp -> user includes that. That dynamic introspection isn't really..possible per se. 

That's why we do the classifier based approach. We could have a tool that inspects the features of various OSes and prints out what the user could include, but that's as good as it's going to get. I would welcome suggestions to improve this though.

**Comment by reckart** [comment]
I understand that everything is compiled ahead of time. 

There is a loader for the native libraries. That loader already takes into account two variables which are only available at runtime:
* the operating system
* whether the CPU is 32 or 64 bit

Do you think it is unreasonable to imagine that the loader could take into account additional information obtained from the environment, e.g.:
* AVX support on the CPU
* Availability of GPU/cuda
* Compatibility of the CPU with MLK-DNN

OS type and 32/64bit are part of the package location for the native libraries - so in order to avoid clashes a measure would need be taken to avoid libs for the same OS/32/64 to overwrite each other (naively: represent more variable in the package name). But in principle, at least to me naively at this point, it seems not per-se impossible.

**Comment by reckart** [comment]
@agibsonccc I thought I head read a comment from you somewhere that there was a (recent) change in the loader (JavaCPP) that would allow to bundle avx and non-avx libraries up together and to decide at load time depending on the CPU which one to use... unfortunately, I don't find that comment anymore and don't remember even exactly where it was... maybe just a wishful dream ;)

**Comment by saudet** [comment]
It's possible to do it with JavaCPP and something like cpu_features, yes, see issue https://github.com/eclipse/deeplearning4j/issues/7834#issuecomment-498170278 and https://github.com/bytedeco/javacpp/issues/496.

**Comment by agibsonccc** [comment]
Thanks @saudet  for commenting! I'll look at this closer next week to figure out how we want to deal with this then. This has been a great discussion so far!

**Comment by agibsonccc** [comment]
@reckart  I have a basic build running on centos 6 now: https://github.com/eclipse/deeplearning4j/actions/workflows/build-deploy-linux-x86_64-compat.yml I'll work on getting a compat classifier you can migrate to for your release. I am going to leave the default on newer versions so people have to consciously pick both a libc as well as what optimizations they want. I'll ping you when the new compat builds are up on snapshots. It's not ideal, but it's at least a migration path.

**Comment by agibsonccc** [comment]
@reckart minor update here...it took a bit of work to get github actions + our existing configurations working with a centos container, but most things seem to be building with that. We should be able to get the new compat classifier out this week with a bit more polish. We'll also try to be more explicit about glibc compatibility in our documentation as well. Thanks for highlighting this!

**Comment by agibsonccc** [comment]
@reckart  https://oss.sonatype.org/content/repositories/snapshots/org/nd4j/nd4j-native/1.0.0-SNAPSHOT/ snapshots are now published for compat using centos 7. Could you verify that the compat classifier solves your problem? It should just be adding:
```xml
<dependency>
<groupId>org.nd4j</groupId>
<artifactId>nd4j-native</artifactId>
<version>1.0.0-SNAPSHOT</version>
<classifier>linux-x86_64-compat</classifier>
</dependency>
```

to your pom.xml.

**Comment by reckart** [comment]
Cool! Will do and then report.

**Comment by agibsonccc** [comment]
@reckart  appreciated. If this works, we'll ship this for the release coming out likely by next week (this was one of the blockers I wanted to get sorted out) - the docs will reflect the  new behavior as well. Sorry it's not as seamless as using the old default, but I more wanted people to have a migration path that's semi acceptable, not necessarily for us to be stuck on old glibcs for the bulk of the project.

**Comment by reckart** [comment]
@agibsonccc it looks like there is no "normal" `windows-x86_64` native anymore, only the AVX and onednn variants. Is that intentional?

**Comment by agibsonccc** [comment]
@reckart that's actually due to sonatype purging snapshots every 3 days. The builds run every 2 days, but an error snuck in to the latest PR preventing the build. I would just use the compat classifier for now and wait for the builds to rerun.

**Comment by reckart** [comment]
Still getting the error. I though it might be related to stale content in the JavaCPP cache and moved that to the target folder of my project where it gets cleaned up as part of the build... but... still no cigar:

```
Caused by: java.lang.RuntimeException: ND4J is probably missing dependencies. For more information, please refer to: https://deeplearning4j.konduit.ai/nd4j/backend
Caused by: java.lang.UnsatisfiedLinkError: no jnind4jcpu in java.library.path
Caused by: java.lang.UnsatisfiedLinkError: [...]target/javacpp-cache/nd4j-native-1.0.0-SNAPSHOT-linux-x86_64.jar/org/nd4j/nativeblas/linux-x86_64/libjnind4jcpu.so: /lib64/libstdc++.so.6: version `CXXABI_1.3.8' not found (required by [...]target/javacpp-cache/nd4j-native-1.0.0-SNAPSHOT-linux-x86_64.jar/org/nd4j/nativeblas/linux-x86_64/libjnind4jcpu.so)
```

**Comment by reckart** [comment]
@agibsonccc Note that the path seems to be wrong. JavaCPP tries to load from

```
nd4j-native-1.0.0-SNAPSHOT-linux-x86_64.jar/org/nd4j/nativeblas/linux-x86_64/libjnind4jcpu.so
```

but the correct location should probably be

```
nd4j-native-1.0.0-SNAPSHOT-linux-x86_64-compat.jar/org/nd4j/nativeblas/linux-x86_64-compat/libjnind4jcpu.so
```

At least if I look at the compat JAR, the package includes a "compat" in its name.

**Comment by agibsonccc** [comment]
@reckart I suspect it's due to me forgetting to add the classifier to the list for the presets: https://github.com/eclipse/deeplearning4j/pull/9373 I'll ping you after we get snapshots for that. Thanks for being patient with this.

**Comment by reckart** [comment]
@agibsonccc is it correct that this issue is already closed?

**Comment by agibsonccc** [comment]
@reckart thanks for checking, snapshots should be up for you now. If you still have issues, let me know. I recently did a staging repository for compat for testing and that worked as well.

**Comment by reckart** [comment]
I am trying to test this again, but the compat artifacts seem to no longer be available in the SNAPSHOT repo: https://oss.sonatype.org/content/repositories/snapshots/org/nd4j/nd4j-native/1.0.0-SNAPSHOT/

**Comment by reckart** [comment]
Can you point me to the staging repo?

**Comment by agibsonccc** [comment]
M1.1 will be out to maven central as soon as sonatype syncs. Please keep an
eye out for central.

**Comment by reckart** [comment]
Did M1.1 get stuck somewhere?

**Comment by agibsonccc** [comment]
@reckart mainly on timed out windows builds failing. Sonatype only told me to re upload and see how it goes. I didn't really get much help on recovering the build that was already there. They never even gave a reason why it failed in the first place. No code has changed, it's literally just getting the builds running and trying it again and seeing what happens. The windows builds take 5.5 hours and sometimes they fail for timeout reasons, out of memory,.. god knows what else. Sorry for the delay.

**Comment by reckart** [comment]
For upload Timeouts the retry setting in the maven deploy plugin is wonderful. It once allowed me to complete a release on the flaky internet of a train 🙂

**Comment by agibsonccc** [comment]
@reckart nah build/github time outs, not maven. I have all the maven wagon configurations setup already to be more robust to those kinds of failures. Thanks for the tip though! The workflows are here if you're curious: https://github.com/eclipse/deeplearning4j/tree/master/.github/workflows

**Comment by nadre** [comment]
Any updates? :)

**Comment by agibsonccc** [comment]
The bottleneck is sonatype here unfortunately. Please see the linked issue
on the ossrh JIRA.

**Comment by agibsonccc** [comment]
@nadre @reckart 1.0.0-M1.1 is out on maven central. Sorry for the delay!

**Comment by reckart** [comment]
Looks like the linux compat libs work nicely :) The only thing: there don't seem to be any "normal" linux natives anymore - only compat, onednn, etc. but nothing without a suffix.

**Comment by agibsonccc** [comment]
@reckart  ah noticed that thanks! I just kicked another release off and will add that classifier as well. It appears that build failed while the other ones passed.

Now that we've done 1 release I will be able to automate the process now. 

The first manual one was to understand how pushing github actions limits (eg: the 6 hour builds + cuda requiring 2 threads)  - overall the process is repeatable enough now + snapshots have been working fine at this point that I think we can avoid this in the future.
I'll also try to see if there's  a way to automate checking for all of the right artifacts being present. That's not currently  in the builds without some manual work on my part.

 Thanks for the feedback! It's been really helpful.

**Comment by agibsonccc** [comment]
@reckart  should be there now  - https://repo1.maven.org/maven2/org/nd4j/nd4j-native/1.0.0-M1.1/ thanks again for highlighting this!
